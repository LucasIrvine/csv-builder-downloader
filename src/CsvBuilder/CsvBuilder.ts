import {
  COMMA,
  CSV_ENCODING_TYPE,
  DEFAULT_FILENAME,
  DEFAULT_REGEX,
  DEFAULT_SUFFIX,
  EMPTY_STRING,
  NEW_LINE,
  REMOVE_COMMA_REGEX,
  ROW_ARRAY_TYPE,
  CELL_TYPE,
} from '../constants';

import { isString, isArray, isNumber } from '../utils';

export default class CsvBuilder {
  encodingType: string;
  file: string;
  fileSuffix: string;
  filename: string;
  filenameFormatter: Function | boolean;
  includeTimeStamp: boolean;
  nonValueIndices: number[];
  sanitizeRegex: RegExp;
  sanitizeValues: boolean;

  constructor(fileConfig: {
    encodingType?: string;
    file?: string;
    fileSuffix?: string;
    filename?: string;
    filenameFormatter?: Function | boolean;
    includeTimeStamp?: boolean;
    nonValueIndices?: number[];
    sanitizeRegex?: RegExp;
    sanitizeValues?: boolean;
  }) {
    const {
      encodingType = CSV_ENCODING_TYPE,
      file = EMPTY_STRING,
      fileSuffix = DEFAULT_SUFFIX,
      filename = DEFAULT_FILENAME,
      filenameFormatter = false,
      includeTimeStamp = true,
      nonValueIndices = [],
      sanitizeRegex = DEFAULT_REGEX,
      sanitizeValues = true,
    } = fileConfig;

    this.encodingType = encodingType;
    this.file = file;
    this.fileSuffix = fileSuffix;
    this.filename = filename;
    this.filenameFormatter = filenameFormatter;
    this.includeTimeStamp = includeTimeStamp;
    this.nonValueIndices = nonValueIndices;
    this.sanitizeRegex = sanitizeRegex;
    this.sanitizeValues = sanitizeValues;
  }

  static removeCommas(val: string) {
    return val?.toString().replace(REMOVE_COMMA_REGEX, EMPTY_STRING) || EMPTY_STRING;
  }

  sanitize(val: CELL_TYPE, commaOnly?: boolean) {
    if (commaOnly) {
      return CsvBuilder.removeCommas(val.toString());
    }

    return val.toString().replace(this.sanitizeRegex, EMPTY_STRING);
  }

  appendToFile(val: CELL_TYPE, newLine?: boolean) {
    this.file += val;
    newLine && this.addNewLine();
  }

  addCell(val: string, sanitize: boolean = this.sanitizeValues) {
    const value = `${sanitize ? this.sanitize(val) : val}${COMMA}`;

    this.appendToFile(value);

    return this;
  }

  addRow(val: CELL_TYPE | ROW_ARRAY_TYPE, bypassSanitize?: boolean) {
    if (isString(val)) {
      this.appendToFile(val as string, true);
    }

    if (isArray(val)) {
      this.addRowArray(val as ROW_ARRAY_TYPE, bypassSanitize);
    }

    return this;
  }

  addRows(rows: ROW_ARRAY_TYPE[]) {
    isArray(rows) && rows.forEach((row) => this.addRow(row));

    return this;
  }

  addRowArray(rowArray: ROW_ARRAY_TYPE, bypassSanitize?: boolean) {
    if (!isArray(rowArray)) {
      return;
    }

    if (this.sanitizeValues && !bypassSanitize) {
      rowArray = rowArray.map((cell: CELL_TYPE, i: number) => {
        return this.nonValueIndices.indexOf(i) > -1 ? this.sanitize(cell, true) : this.sanitize(cell);
      });
    }

    this.appendToFile(rowArray.join(COMMA), true);
  }

  addNewLine(count: number = 1) {
    if (count < 1 || !isNumber(count)) {
      return this;
    }

    while (count-- as number) {
      this.appendToFile(NEW_LINE);
    }

    return this;
  }

  addSection(title: string, headers: ROW_ARRAY_TYPE, rows: ROW_ARRAY_TYPE[], newLines: number) {
    if (isString(title)) {
      this.addRow(title);
    }

    if (isArray(headers)) {
      this.addRow(headers, true);
    }

    if (isArray(rows) && isArray(rows[0])) {
      this.addRows(rows);
    }

    this.addNewLine(newLines);

    return this;
  }

  getEncodedFile() {
    return `${this.encodingType}${encodeURI(this.file)}`;
  }

  getFilename() {
    if (this.filenameFormatter && typeof this.filenameFormatter === 'function') {
      const filename = this.filenameFormatter.call(this);

      return (isString(filename) && filename) || DEFAULT_FILENAME;
    }

    if (this.includeTimeStamp) {
      return `${this.filename}-${new Date().valueOf()}${this.fileSuffix}`;
    }

    return `${this.filename}${this.fileSuffix}`;
  }
}
