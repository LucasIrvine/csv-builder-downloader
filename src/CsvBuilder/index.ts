import CsvBuilder from './CsvBuilder';

export default CsvBuilder;
