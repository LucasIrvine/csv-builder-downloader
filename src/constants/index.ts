// Types
export type CELL_TYPE = string | number | boolean;
export type ROW_ARRAY_TYPE = (string | number | boolean)[];
export type SECTION_TYPE = { title: string; headers: ROW_ARRAY_TYPE; rows: ROW_ARRAY_TYPE[]; newLines: number };
export type CSV_BUILDER_CONFIG_TYPE = {
  encodingType?: string;
  file?: string;
  fileSuffix?: string;
  filename?: string;
  includeTimeStamp?: boolean;
  nonValueIndices?: number[];
  sanitizeRegex?: RegExp;
  sanitizeValues?: boolean;
};
// Params
export const MAX_NEW_LINES = 20;

// Regex
export const ALPHA_NUMERIC_DATE_REGEX = /[^a-zA-Z0-9:\./_\-@$ ]/g;
export const REMOVE_COMMA_REGEX = /,/g;

// Values
export const NEW_LINE = '\n';
export const COMMA = ',';
export const EMPTY_STRING = '';
export const ZERO = 0;
export const DECIMAL_ZERO = 0.0;
export const EMPTY_ARRAY = [];
export const EMPTY_OBJECT = {};
export const NA = 'N/A';

// Files
export const CSV_ENCODING_TYPE = `data:text/csv;charset=utf-8${COMMA}`;
export const DEFAULT_FILENAME = 'csv-download';
export const DEFAULT_SUFFIX = '.csv';

// Default type values
export const STRING_DEFAULT = EMPTY_STRING;
export const NUMBER_DEFAULT = ZERO;
export const ARRAY_DEFAULT = EMPTY_ARRAY;
export const OBJECT_DEFAULT = EMPTY_OBJECT;
export const NULL_DEFAULT = EMPTY_STRING;
export const DEFAULT_REGEX = ALPHA_NUMERIC_DATE_REGEX;
