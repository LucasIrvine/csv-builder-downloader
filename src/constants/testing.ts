import {
  COMMA,
  CSV_ENCODING_TYPE,
  EMPTY_STRING,
  DEFAULT_FILENAME,
  DEFAULT_SUFFIX,
  DEFAULT_REGEX,
  REMOVE_COMMA_REGEX,
} from '.';

// These are excluded from the ts build
export const CSV_BUILDER_CONFIG_SIMPLE = {
  filename: 'test-file-simple',
};
export const CHARS = `abcXYZ123,,..!!@@##$$%%^^&&**(())__++--==~~[[]]\\{{}}||;;''::""//<<>>??¡™£¢∞§¶•ªº–≠œ∑´®†¥¨ˆøπ“‘«åß∂ƒ©˙∆˚¬…æ≈ç√∫˜µ≤≥÷`;
export const DEFAULT_ALLOWED_CHARS = 'abcXYZ123..@@$$__--:://';
export const STRING_INPUT_1 = 'Is this your homework larry?';
export const STRING_INPUT_2 = 'Can you not';
export const ENCODED_STRING_INPUT = encodeURI(STRING_INPUT_2);
export const NUMERIC_INPUT = 1000000.713;
export const COMMA_INPUT = 'Well, well, well...';
export const NO_COMMA_INPUT = 'Well well well...';
export const FORMATTED_NUMERIC_INPUT = '$1,000,000.713';
export const DATE_1 = '01/04/85';
export const DATE_2 = '12-05-1983';
export const TEST_ROW = [STRING_INPUT_1, NUMERIC_INPUT, STRING_INPUT_2, FORMATTED_NUMERIC_INPUT, DATE_1, DATE_2, CHARS];
export const TEST_ROWS_ARRAY = [TEST_ROW, TEST_ROW, TEST_ROW];
export const ENCODED_FILE = `${CSV_ENCODING_TYPE}${ENCODED_STRING_INPUT}${COMMA}`;
export const MOCK_FILENAME = 'MyCsvDownload';
export const MOCK_FUNCTION = () => MOCK_FILENAME;
export const MOCK_BOOLEAN = true;
export const MOCK_OBJECT = { test: 'object' };
export const ALL_OPTIONS_DEFAULT = {
  encodingType: CSV_ENCODING_TYPE,
  file: EMPTY_STRING,
  fileSuffix: DEFAULT_SUFFIX,
  filename: DEFAULT_FILENAME,
  includeTimeStamp: true,
  nonValueIndices: [],
  sanitizeRegex: DEFAULT_REGEX,
  sanitizeValues: true,
};
export const ALL_OPTIONS_CUSTOM = {
  encodingType: 'data:text/csv;charset=utf-16',
  file: 'Already, Have, Header, Data,',
  fileSuffix: '.xlxs',
  filename: 'Not_A_Very_Good_Name',
  includeTimeStamp: false,
  nonValueIndices: [0, 1],
  sanitizeRegex: REMOVE_COMMA_REGEX,
  sanitizeValues: false,
};
