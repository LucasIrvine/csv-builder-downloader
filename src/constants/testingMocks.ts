import { COMMA, CSV_ENCODING_TYPE } from '../constants';

// These are excluded from the ts build
export const CHARS = `abcXYZ123,,..!!@@##$$%%^^&&**(())__++--==~~[[]]\\{{}}||;;''::""//<<>>??¡™£¢∞§¶•ªº–≠œ∑´®†¥¨ˆøπ“‘«åß∂ƒ©˙∆˚¬…æ≈ç√∫˜µ≤≥÷`;
export const DEFAULT_ALLOWED_CHARS = 'abcXYZ123..@@$$__--:://';
export const STRING_INPUT_1 = 'Is this your homework larry?';
export const STRING_INPUT_2 = 'Can you not';
export const ENCODED_STRING_INPUT = encodeURI(STRING_INPUT_2);
export const NUMERIC_INPUT = 1000000.713;
export const FORMATTED_NUMERIC_INPUT = '$1,000,000.713';
export const DATE_1 = '01/04/85';
export const DATE_2 = '12-05-1983';
export const TEST_ROW = [STRING_INPUT_1, NUMERIC_INPUT, STRING_INPUT_2, FORMATTED_NUMERIC_INPUT, DATE_1, DATE_2, CHARS];
export const TEST_ROWS_ARRAY = [TEST_ROW, TEST_ROW, TEST_ROW];
export const ENCODED_FILE = `${CSV_ENCODING_TYPE}${ENCODED_STRING_INPUT}${COMMA}`;
export const MOCK_FILENAME = 'MyCsvDownload';
