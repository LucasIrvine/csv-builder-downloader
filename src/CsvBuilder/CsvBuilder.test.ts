import CsvBuilder from './index';

import { COMMA, CSV_BUILDER_CONFIG_SIMPLE, CSV_ENCODING_TYPE, EMPTY_STRING, NEW_LINE } from '../constants';

// Keep in file so it doesnt get packaged in prod build with other constants
const CHARS = `abcXYZ123,,..!!@@##$$%%^^&&**(())__++--==~~[[]]\\{{}}||;;''::""//<<>>??¡™£¢∞§¶•ªº–≠œ∑´®†¥¨ˆøπ“‘«åß∂ƒ©˙∆˚¬…æ≈ç√∫˜µ≤≥÷`;
const DEFAULT_ALLOWED_CHARS = 'abcXYZ123..@@$$__--:://';
const STRING_INPUT_1 = 'Is this your homework larry?';
const STRING_INPUT_2 = 'Can you not';
const ENCODED_STRING_INPUT = encodeURI(STRING_INPUT_2);
const NUMERIC_INPUT = 1000000.713;
const FORMATTED_NUMERIC_INPUT = '$1,000,000.713';
const DATE_1 = '01/04/85';
const DATE_2 = '12-05-1983';
const TEST_ROW = [STRING_INPUT_1, NUMERIC_INPUT, STRING_INPUT_2, FORMATTED_NUMERIC_INPUT, DATE_1, DATE_2, CHARS];
const TEST_ROWS_ARRAY = [TEST_ROW, TEST_ROW, TEST_ROW];

describe('CsvBuilder', () => {
  let csvBuilder: CsvBuilder;

  beforeEach(() => {
    csvBuilder = new CsvBuilder(CSV_BUILDER_CONFIG_SIMPLE);
  });

  describe('constructor', () => {
    it('should set the config values correctly.', () => {
      expect(csvBuilder.filename).toEqual(CSV_BUILDER_CONFIG_SIMPLE.filename);
    });
  });

  describe('removeCommas (static)', () => {
    it('should remove all commas from a value', () => {
      const input = '1,000,000,000';
      const expectedOutput = '1000000000';
      expect(CsvBuilder.removeCommas(input)).toEqual(expectedOutput);
    });
  });

  describe('.sanitize', () => {
    it('should allow only characters in the default regex (alphanumeric, dates)', () => {
      expect(csvBuilder.sanitize(CHARS)).toEqual(DEFAULT_ALLOWED_CHARS);
      expect(csvBuilder.sanitize(CHARS)).toMatchSnapshot();
    });
  });

  describe('.appendToFile', () => {
    it('should append to current file string', () => {
      const expectedOutput1 = `${STRING_INPUT_1}`;
      const expectedOutput2 = `${STRING_INPUT_1}${STRING_INPUT_2}`;

      csvBuilder.appendToFile(STRING_INPUT_1);
      expect(csvBuilder.file).toEqual(expectedOutput1);

      csvBuilder.appendToFile(STRING_INPUT_2);
      expect(csvBuilder.file).toEqual(expectedOutput2);

      expect(csvBuilder.file).toMatchSnapshot();
    });
  });

  describe('.addCell', () => {
    it('should produce the same file, given the same input', () => {
      csvBuilder.addCell(STRING_INPUT_1);

      expect(csvBuilder.file).toMatchSnapshot();
    });
  });

  describe('.addRow', () => {
    it('should produce the same file, given the same `string` input', () => {
      csvBuilder.addRow(STRING_INPUT_1);

      expect(csvBuilder.file).toMatchSnapshot();
    });

    it('should produce the same file, given the same `array` input', () => {
      csvBuilder.addRow(TEST_ROW);

      expect(csvBuilder.file).toMatchSnapshot();
    });

    it('should only return the class instance if neither a `string` or `array` is passed', () => {
      // @ts-ignore: passing in the wrong type
      expect(csvBuilder.addRow(true)).toBe(csvBuilder);
      expect(csvBuilder.file).toEqual(EMPTY_STRING);
    });
  });

  describe('.addRows', () => {
    it('should produce the same file, given the same input', () => {
      csvBuilder.addRows(TEST_ROWS_ARRAY);

      expect(csvBuilder.file).toMatchSnapshot();
    });

    it('should call .addRow for each item in the array', () => {
      const arrayLength = TEST_ROWS_ARRAY.length;

      csvBuilder.addRow = jest.fn();
      csvBuilder.addRows(TEST_ROWS_ARRAY);

      expect(csvBuilder.addRow).toHaveBeenCalledTimes(arrayLength);
    });

    it('should only return the class instance if anything but an `array` is passed', () => {
      // @ts-ignore: passing in the wrong type
      expect(csvBuilder.addRows(true)).toBe(csvBuilder);
      expect(csvBuilder.file).toEqual(EMPTY_STRING);
    });
  });

  describe('.addRowArray', () => {
    it('should produce the same file, given the same input', () => {
      csvBuilder.addRowArray(TEST_ROW);

      expect(csvBuilder.file).toMatchSnapshot();
    });

    it('should return early and not add to the file if a non-array value is passed', () => {
      // @ts-ignore: passing in the wrong type
      csvBuilder.addRowArray(STRING_INPUT_1);

      expect(csvBuilder.file).toEqual(EMPTY_STRING);
    });
  });

  describe('.addNewLine', () => {
    it('should not add a new line with a falsy value passed', () => {
      // @ts-ignore: passing in the wrong type
      csvBuilder.addNewLine(null);

      expect(csvBuilder.file).toEqual(EMPTY_STRING);
    });

    it('should not add a new line with a non-numerical value passed', () => {
      // @ts-ignore: passing in the wrong type
      csvBuilder.addNewLine(STRING_INPUT_2);

      expect(csvBuilder.file).toEqual(EMPTY_STRING);
    });

    it('should default to one new line when no count value passed', () => {
      csvBuilder.addNewLine();

      expect(csvBuilder.file).toEqual(NEW_LINE);
    });
  });

  describe('.getEncodedFile', () => {
    it('should produce the same file, given the same input', () => {
      csvBuilder.addCell(STRING_INPUT_1);
      expect(csvBuilder.getEncodedFile()).toMatchSnapshot();
    });

    it("should create the encoded file from the CsvBuilder's default encoding type (CSV_ENCODING_TYPE) and the current file uri encoded", () => {
      const expectedOutput = `${CSV_ENCODING_TYPE}${ENCODED_STRING_INPUT}${COMMA}`;

      csvBuilder.addCell(STRING_INPUT_2);
      expect(csvBuilder.getEncodedFile()).toEqual(expectedOutput);
    });
  });

  // TODO
  describe('.addSection', () => {
    it('', () => {});
  });
});
