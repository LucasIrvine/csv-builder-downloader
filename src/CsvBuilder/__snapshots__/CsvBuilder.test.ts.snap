// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvBuilder .addCell should produce the same file, given the same input 1`] = `"Is this your homework larry,"`;

exports[`CsvBuilder .addRow should produce the same file, given the same \`array\` input 1`] = `
"Is this your homework larry,1000000.713,Can you not,$1000000.713,01/04/85,12-05-1983,abcXYZ123..@@$$__--:://
"
`;

exports[`CsvBuilder .addRow should produce the same file, given the same \`string\` input 1`] = `
"Is this your homework larry?
"
`;

exports[`CsvBuilder .addRowArray should produce the same file, given the same input 1`] = `
"Is this your homework larry,1000000.713,Can you not,$1000000.713,01/04/85,12-05-1983,abcXYZ123..@@$$__--:://
"
`;

exports[`CsvBuilder .addRows should produce the same file, given the same input 1`] = `
"Is this your homework larry,1000000.713,Can you not,$1000000.713,01/04/85,12-05-1983,abcXYZ123..@@$$__--:://
Is this your homework larry,1000000.713,Can you not,$1000000.713,01/04/85,12-05-1983,abcXYZ123..@@$$__--:://
Is this your homework larry,1000000.713,Can you not,$1000000.713,01/04/85,12-05-1983,abcXYZ123..@@$$__--:://
"
`;

exports[`CsvBuilder .appendToFile should append to current file string 1`] = `"Is this your homework larry?Can you not"`;

exports[`CsvBuilder .getEncodedFile should produce the same file, given the same input 1`] = `"data:text/csv;charset=utf-8,Is%20this%20your%20homework%20larry,"`;

exports[`CsvBuilder .sanitize should allow only characters in the default regex (alphanumeric, dates) 1`] = `"abcXYZ123..@@$$__--:://"`;
